apps/internal/base/base.go

```diff
@@ -417,103 +400,66 @@ func (b Client) AcquireTokenOnBehalfOf(ctx context.Context, onBehalfOfParams Acq
+	var m manager = b.manager
 	if authParams.AuthorizationType == authority.ATOnBehalfOf {
-		if s, ok := b.pmanager.(cache.Serializer); ok {
-			suggestedCacheKey := token.CacheKey(authParams)
-			err = b.cacheAccessor.Replace(ctx, s, cache.ReplaceHints{PartitionKey: suggestedCacheKey})
-			if err != nil {
-				return ar, err
-			}
-			defer func() {
-				err = b.export(ctx, s, suggestedCacheKey, err)
-			}()
-		}
-		account, err = b.pmanager.Write(authParams, token)
-		if err != nil {
-			return ar, err
-		}
-	} else {
-		if s, ok := b.manager.(cache.Serializer); ok {
-			suggestedCacheKey := token.CacheKey(authParams)
-			err = b.cacheAccessor.Replace(ctx, s, cache.ReplaceHints{PartitionKey: suggestedCacheKey})
-			if err != nil {
-				return ar, err
-			}
-			defer func() {
-				err = b.export(ctx, s, suggestedCacheKey, err)
-			}()
-		}
-		account, err = b.manager.Write(authParams, token)
-		if err != nil {
-			return ar, err
-		}
+		m = b.pmanager
+	}
+	ar := AuthResult{}
+	key := token.CacheKey(authParams)
+	err := b.cacheAccessor.Replace(ctx, m, cache.ReplaceHints{PartitionKey: key})
+	if err != nil {
+		return ar, err
+	}
+	account, err := m.Write(authParams, token)
+	if err != nil {
+		return ar, err
+	}
+	if ar, err = NewAuthResult(token, account); err == nil {
+		err = b.cacheAccessor.Export(ctx, b.manager, cache.ExportHints{PartitionKey: key})
 	}
-	ar, err = NewAuthResult(token, account)
 	return ar, err
 }
 
-func (b Client) AllAccounts(ctx context.Context) (accts []shared.Account, err error) {
-	if s, ok := b.manager.(cache.Serializer); ok {
-		suggestedCacheKey := b.AuthParams.CacheKey(false)
-		err = b.cacheAccessor.Replace(ctx, s, cache.ReplaceHints{PartitionKey: suggestedCacheKey})
-		if err != nil {
-			return accts, err
-		}
-		defer func() {
-			err = b.export(ctx, s, suggestedCacheKey, err)
-		}()
+func (b Client) AllAccounts(ctx context.Context) ([]shared.Account, error) {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	key := b.AuthParams.CacheKey(false)
+	err := b.cacheAccessor.Replace(ctx, b.manager, cache.ReplaceHints{PartitionKey: key})
+	if err != nil {
+		return nil, err
 	}
-
-	accts = b.manager.AllAccounts()
-	return accts, err
+	return b.manager.AllAccounts(), nil
 }
 
-func (b Client) Account(ctx context.Context, homeAccountID string) (acct shared.Account, err error) {
+func (b Client) Account(ctx context.Context, homeAccountID string) (shared.Account, error) {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
 	authParams := b.AuthParams // This is a copy, as we dont' have a pointer receiver and .AuthParams is not a pointer.
 	authParams.AuthorizationType = authority.AccountByID
 	authParams.HomeAccountID = homeAccountID
-	if s, ok := b.manager.(cache.Serializer); ok {
-		suggestedCacheKey := b.AuthParams.CacheKey(false)
-		err = b.cacheAccessor.Replace(ctx, s, cache.ReplaceHints{PartitionKey: suggestedCacheKey})
-		if err != nil {
-			return acct, err
-		}
-		defer func() {
-			err = b.export(ctx, s, suggestedCacheKey, err)
```

Comment:
Does this call to `export()` need to be preserved?

Reply:
No; export writes the in-memory cache to the persistence implementation. This method copies persisted data into the in-memory cache and searches it for an account. It doesn't create any data, so it has nothing to add to the persisted data.